Save intermediate state when restoring the Byron chain in the stake pools worker

hasufell opened the pull request:
This completely skips byron era for mainnet when syncing stake pools.

I extracted the first shelley block by looking at a fully synced stake pool DB and selecting the lowest `block_height` from the `pool_production` table.

For non-mainnet it will record the last byron block of a chunk from the node into the `byron_headers` table and read it out during `readPoolProductionCursor` if there are no stake pools and otherwise ignore it.

Remarks:

- the "genesis shelley header" is hardcoded, because the node currently does not have an endpoint to get this

Commit:
Sync from Shelley hard fork on mainnet wrt #1963

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,8 +494,24 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+                (either (error . show) id $
+                    fromText "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
+                (either (error . show) id $
+                    fromText "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
```
use instead: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Unsafe.hs#L90

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,8 +494,24 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+        fromDB <- atomically $ readPoolProductionCursor (max 100 k)
+        case fromDB of
+            [] -> pure [genesisShelleyHeader]
+            xs -> pure xs
```
Note that this is (almost) equivalent to: `genesisShelleyHeader:xs`. When looking for an intersection, the node look at intersection points in order. The last item of that list is considered the "most interesting" point.

Commit:
Record headers in stake pools DB and use them as sync point wrt #1963

This will allow to cancel syncing during byron era
for all networks and pick up where we left off.

We use the same method for shelley as well instead of
using stake pool production table to unify behavior.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -181,6 +188,18 @@ mPutPoolProduction point poolId db@PoolDatabase{pools} =
+mPutByronHeader :: BlockHeader -> ModelPoolOp ()
+mPutByronHeader point db@PoolDatabase{byronHeaders} =
+    ( Right ()
+    , db { byronHeaders = point:byronHeaders })
```
Hi @hasufell.

We revised the `Pool.DB.Model` module quite a bit in https://github.com/input-output-hk/cardano-wallet/pull/2132.

So you'd now probably need to write something like this:
```suggestion
mPutByronHeader :: BlockHeader -> ModelOp ()      
mPutByronHeader header = modify #byronHeaders (header :)  
```

Also, I noticed that the model implementation don't check to see if the point is already in the list. Should we check? We do appear check for this in the SQLite implementation:

```hs
        putByronHeader point = ExceptT $
            handleConstraint (ErrPointAlreadyExists point) $
                let record = mkByronHeader point
                    key = byronBlockHeight record
                in repsert (ByronHeadersKey key) record
```

hasufell replied:
`repsert` will overwrite, so the catching here seems unnecessary

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -181,6 +188,18 @@ mPutPoolProduction point poolId db@PoolDatabase{pools} =
+
+mListByronHeaders :: Int -> PoolDatabase -> ([BlockHeader], PoolDatabase)
+mListByronHeaders k db@PoolDatabase{byronHeaders} =
+    (take k byronHeaders, db)
```
With the revised `ModelOp` type, we can write the function like this:

```suggestion
mListByronHeaders :: Int -> ModelOp [BlockHeader]     
mListByronHeaders k = take k <$> get #byronHeaders
```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -181,6 +188,18 @@ mPutPoolProduction point poolId db@PoolDatabase{pools} =
+    (take k byronHeaders, db)
+
+mDropByronHeaders :: ModelPoolOp ()
+mDropByronHeaders db = (Right (), db { byronHeaders = [] })
```
With the revised `ModelOp` type, we can write this function like this:

```suggestion
mDropByronHeaders :: ModelOp ()
mDropByronHeaders = modify #byronHeaders (const [])
```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +560,23 @@ newDBLayer trace fp timeInterpreter = do
+
+        dropByronHeaders =
+            let query = [s|DELETE FROM byron_headers;|]
+            in rawExecute query []
```
This SQL query looks perfectly fine to me, but I wonder if  `persistent` has an equivalent that would mean we could avoid SQL here?

hasufell replied:
I couldn't find it.

KtorZ replied:
`deleteWhere []`

KtorZ replied:
with a proper type annotation on `[]` to let persistent know what table.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -468,9 +469,14 @@ newDBLayer trace fp timeInterpreter = do
                 [Desc PoolProductionSlot, LimitTo k]
+            case pp of
+                -- byron headers might not be inserted
+                -- continously, so only pick the last
```
Small typo:
```suggestion
                -- continuously, so only pick the last.
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,8 +498,22 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+                (Hash $ unsafeFromHex
+                    "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
+                (Hash $ unsafeFromHex
+                    "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
```
A couple of suggestions:

1. Move this definition to the top-level, and change the comment to a Haddock-style comment.
2. Make the name of the constant clearly indicate that it's for `mainnet`. So, perhaps something like `mainnetGenesisShelleyHeader` or `mainnetShelleyGenesisHeader`?

Anviking replied:
I'm not sure the name "genesis" here actually makes sense. I think the Shelley genesis file is used for the hard-fork, but doesn't actually correspond to the first Shelley block.

Even so, wouldn't you want `finalByronBlockHeader` instead, such that the node gives you the first Shelley block and not the second?

hasufell replied:
@jonathanknowles 

> Move this definition to the top-level, and change the comment to a Haddock-style comment.

What would be the reason for that? It's only used once. I usually prefer to not pollute the module namespace, unless it's really a shared entity.

> Make the name of the constant clearly indicate that it's for

What about `firstShelleyBlockMainnet`?

----

@Anviking 

> Even so, wouldn't you want finalByronBlockHeader instead, such that the node gives you the first Shelley block and not the second?

You're absolutely right. I double checked this in the sqlite database and we need the last byron header indeed.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -506,24 +524,35 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+                BlockByron blk -> do
+                  dropByronHeaders
+                  handleErr (putByronHeader (toByronBlockHeader gp blk))
+                BlockShelley _ -> pure ()
```
**Suggestion**: Perhaps break up this section a bit, to reduce the level of nesting?

Something like:

```hs
atomically $
    processBlocksByron blocks
    processBlocksShelley blocks
```
And then define:
```hs
processBlocksShelley blocks =
    forM_ blocks $ \case
    BlockByron _ ->
        ...
    BlockShelley blk -> do
        -- Record pool certificates
        ...

processBlocksByron blocks = 
    case NE.last blocks of
        BlockByron blk ->
            dropByronHeaders
                handleErr (putByronHeader (toByronBlockHeader gp blk))
        BlockShelley _ -> 
            ...
```

hasufell replied:
Though about it. But I think it might tempt someone to use two forM blocks by accident, which would be a performance penalty. This way we can see that we want to do as little processing as possible.

hasufell replied:
In fact, we're already traversing twice due to `NE.last`... I'll try to fix that

hasufell replied:
```hs
        -- | Like 'forM_', except runs the second action for the last element as
        -- well (in addition to the first action).
        forML_ :: (Monad m)
            => [a]
            -> (a -> m b) -- ^ action to run for all elements
            -> (a -> m c) -- ^ action to run for the last element
            -> m ()
        forML_ []  _ _ = pure ()
        forML_ [x] a1 a2 = a1 x >> a2 x >> forML_ [] a1 a2
        forML_ (x:xs) a1 a2 = a1 x >> forML_ xs a1 a2
```

hasufell replied:
Surprisingly, this made it slower... maybe the other form has better list fusion/inlining and memoization.

Anviking replied:
What if you write the _first_ byron block header instead of the _last_, does that help?

The checkpoint will be a bit (1000 blocks?) old, but I can't see how that would matter.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/MVar.hs:
```diff
@@ -151,6 +154,16 @@ newDBLayer timeInterpreter = do
+                (alterPoolDB errPointAlreadyExists db $ mPutByronHeader point)
+
+        listByronHeaders k =
+            modifyMVar db (pure . swap . mListByronHeaders k)
```
Request: use `readPoolDB` if possible, as it's a read-only operation.

```suggestion
        listByronHeaders =
            readPoolDB db . mListByronHeaders
```

jonathanknowles reviewed (changes requested):
Hi @hasufell 

This looks good so far. :+1: 

I have a few suggestions (see review comments), and a couple of requests:

- Use `readPoolDB` within `listByronHeaders` (https://github.com/input-output-hk/cardano-wallet/pull/2138#discussion_r488404944).
- We should probably add some **property tests** for the newly-added DB operations. We're currently in the middle of [adjusting](https://github.com/input-output-hk/cardano-wallet/pull/2136) our [pool DB property test suite](https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Cardano/Pool/DB/Properties.hs) so that it can test both the model and SQLite implementations. So whatever properties we write should pass with both implementations.

Also, we recently changed the type of `ModelOp` for the pool database. So there will be some changes necessary here. However, I've added comments to the PR showing how to make these changes. Let me know if I can help in any way!

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,8 +498,22 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
       where k = fromIntegral $ getQuantity getEpochStability
+            -- We only have stake pools since shelley era.
+            -- This is hardcoded for mainnet and will do nothing for other
+            -- networks.
```
Neat!

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,8 +498,22 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            -- We only have stake pools since shelley era.
+            -- This is hardcoded for mainnet and will do nothing for other
+            -- networks.
+            -- TODO: query cardano-node for hard fork block (not yet possible)
```
> not **yet** possible

Is there something suggesting that it _will_ be?

hasufell replied:
My intention to file a feature request!

KtorZ replied:
Let's open the ticket first, and reference it from the TODO comment.

hasufell replied:
https://jira.iohk.io/browse/CAD-1850

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +559,20 @@ newDBLayer trace fp timeInterpreter = do
+                , LimitTo k
+                ]
+
+        dropByronHeaders = deleteWhere @_ @_ @ByronHeaders []
```
:+1:

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -837,6 +856,21 @@ fromPoolProduction (PoolProduction pool slot headerH parentH height) =
+fromByronHeaders :: ByronHeaders -> BlockHeader
+fromByronHeaders ByronHeaders { .. } =
+    BlockHeader byronSlot (Quantity byronBlockHeight) (getBlockId byronHeaderHash)
+        (getBlockId byronParentHash)
```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

```suggestion
    BlockHeader byronSlot
        (Quantity byronBlockHeight)
        (getBlockId byronHeaderHash)
        (getBlockId byronParentHash)
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,36 +499,81 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+                (Hash $ unsafeFromHex
+                    "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
+                (Hash $ unsafeFromHex
+                    "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
```
I still think this would be better as a top-level constant.

Embedding this as a nested constant means that it's doesn't fit within our [line limit](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters).

But by moving this to the top-level, we can completely avoid this problem.

As a bonus, we can add a Haddock comment:

```suggestion
-- | The very last block header of the Byron era, just before the transition
--   to Shelley.
-- 
-- This is hard-coded for mainnet and will do nothing for other networks.
--
-- TODO: query cardano-node for hard fork block (not yet possible).
-- https://jira.iohk.io/browse/CAD-1850
--
lastByronBlock :: BlockHeader
lastByronBlock = BlockHeader
    (SlotNo 4492799)
    (Quantity 4490510)
    (Hash $ unsafeFromHex
        "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
    (Hash $ unsafeFromHex
        "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,36 +499,81 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+        forLastBlock = \case
+            BlockByron blk -> do
+                c <- liftIO $ readIORef chunkCounter
+                when (c `mod` 50 == 0) $ do -- every 50k blocks, insert a header
```
Is it possible that we might repeatedly skip over this condition? As far as I know, we don't really control how many blocks we're given by the block following protocol. Suppose we're initially given 10 blocks, and then from that point forward, repeatedly given 50 blocks? Would the `c mod 50` condition not then would always evaluate to `10`?

What if instead we just write a header once per epoch? We already track the _last-seen epoch_ in the `mkLatestGarbageCollectionEpochRef`. Perhaps there is a way to reuse that reference here, and make it into a more general purpose reference of the "last-seen epoch"?

Whenever we get a new set of blocks, we can check whether or not we've transitioned into a new epoch, and if so:

 1. Write a new block header (if we're in the Byron era).
 2. Perform garbage collection of retired pools (if we're in the Shelley era).
 3. Update the latest epoch reference.

KtorZ replied:
What is this even needed? Writing one block header every batch isn't going to be very costly. `forLastBlock` could be simplified to write block headers to the database everytime it is called, there's no need for further filtering.

KtorZ replied:
More importantly, the point of writing those headers is to maintain a list of candidates that'll be used for finding an intersection in case the system restart. So, having gaps of 50k blocks (or even, of an entire epoch) is _not acceptable_. 

The chain follower implementation will yield blocks by batches while syncing far from the tip, and once reaching the "unstable edge" (i.e. k blocks from the network tip), it'll start yielding them in a slower fashion. In practice, we want to have some a high density of blocks near the tip, and can resort to a sparse view for deeper blocks. This is implemented by the `sparseCheckpoints` function in `Cardano.Wallet.DB` which gives you a list of block numbers to preserve given `k` and a current chain height

KtorZ replied:
The approach taken for wallet workers is to always store the last block received from the chain follower and, regularly _prune_ old checkpoints that are no longer required (because, the chain is always going forward). This works well in practice, so I'll suggest to stick to this :+1:

hasufell replied:
I did all those changes... but note that a regular uninterrupted sync on non-mainnet will be slower due to the inserts.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,36 +499,81 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
         -> IO (FollowAction ())
-    forward latestGarbageCollectionEpochRef blocks (_nodeTip, _pparams) = do
-        atomically $ forM_ blocks $ \case
+    forward latestGarbageCollectionEpochRef chunkCounter blocks (_nodeTip, _pparams) = do
```
```suggestion
    forward latestGarbageCollectionEpochRef chunkCounter blocks _ = do
```

(Tangentially: https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters)

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -468,9 +468,14 @@ newDBLayer trace fp timeInterpreter = do
+                -- byron headers might not be inserted
+                -- continuously, so only pick the last.
+                [] -> listByronHeaders 1
+                xs -> pure xs
```
Request: make this a bit shorter. (See [line limit](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters).)

Here's one potentially nice way:
```suggestion
            pp <- map (snd . fromPoolProduction . entityVal) <$> selectList
                []
                [Desc PoolProductionSlot, LimitTo k]
            case pp of
                -- byron headers might not be inserted
                -- continuously, so only pick the last.
                [] -> listByronHeaders 1
                xs -> pure $ reverse xs
```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +559,20 @@ newDBLayer trace fp timeInterpreter = do
+
+        dropByronHeaders = deleteWhere @_ @_ @ByronHeaders []
+
+
```
```suggestion

```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -396,6 +403,15 @@ mRemoveRetiredPools epoch = do
     pure certificates
 
+mPutByronHeader :: BlockHeader -> ModelOp ()      
+mPutByronHeader header = modify #byronHeaders (header :)  
```
:+1: 

This is much more concise!

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -396,6 +403,15 @@ mRemoveRetiredPools epoch = do
+mPutByronHeader header = modify #byronHeaders (header :)  
+
+mListByronHeaders :: Int -> ModelOp [BlockHeader]     
+mListByronHeaders k = take k <$> get #byronHeaders
```
:+1:

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -396,6 +403,15 @@ mRemoveRetiredPools epoch = do
+mListByronHeaders k = take k <$> get #byronHeaders
+
+mDropByronHeaders :: ModelOp ()
+mDropByronHeaders = modify #byronHeaders (const [])
```
:+1:

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/MVar.hs:
```diff
@@ -146,6 +149,15 @@ newDBLayer timeInterpreter = do
+            void . alterPoolDB (const Nothing) db . mPutByronHeader $ point
+
+        listByronHeaders =
+            readPoolDB db . mListByronHeaders
```
:+1:

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -223,6 +223,20 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+        :: Int -- ^ limit
+        -> stm [BlockHeader]
+        -- ^ List byron headers, usually stored during syncing
+        --
```
```suggestion
```

jonathanknowles reviewed (changes requested):
Hi @hasufell 

Many thanks for the update! I see there have been quite a few improvements.

I still have three concerns:

1. Given that we only need to write a block header to the database occasionally, I can't help wondering whether there isn't a simpler way to handle this? We already track transitions across epoch boundaries in order to limit the garbage collection of retired pools: can we not reuse the same technique (and even the same reference) to decide when to write block headers? I'm also concerned that the existing logic might inadvertently fail to fire (though I could be wrong about this). See my comment [here](https://github.com/input-output-hk/cardano-wallet/pull/2138#discussion_r488659342) for more info.

2. There don't seem to be any property tests just yet. I think it would be worth adding some basic property tests for the new DB operations, in the same style as those in `lib/core/test/unit/Cardano/Pool/DB/Properties.hs`.

3. There are some other coding standards issues that should really be fixed. (See comments.) I've added a suggestion for each one, so it should be fast to fix these.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,36 +499,89 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+                liftIO $ modifyIORef' chunkCounter (+1)
+            BlockShelley _ -> pure ()
+
+        {-# INLINE handleErr #-}
```
Leave inlines to the compiler by default. Add a pragma if it appears to be necessary.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -505,25 +511,49 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            -> (a -> m c) -- ^ action to run for the last element
+            -> m ()
+        {-# INLINE forML_ #-}
+        forML_ ne a1 a2 = go (NE.toList ne)
```
**Suggestion:** replace this abstraction with something simpler.

**Justification:**

It seems to me that very short and cryptic names are generally reserved for functions that are very general in nature, such as `forM`, or `scanM` and so on. Since we tend to reuse these general functions a great deal, it doesn't matter that they have short names: the names become part of the lexicon that everyone gets used to over time. And because we have to read them so frequently, being short has an advantage.

But this function is specialized to a single use case. Having a cryptic name means that the reader has to go to the definition of the function to figure out what it's doing.

The overall design also seems a bit brittle: if we need to change the way we record block headers (for example), then it's likely we'll need to replace `forML_` with something else entirely.

**Proposal:**

Why not write something like this, instead:
```hs
forward latestGarbageCollectionEpochRef blocks _ = do
    atomically $ do
        recordPoolCertificates blocks
        recordBlockHeaders blocks
    pure Continue

```
And then just let `recordPoolCertificates` and `recordBlockHeaders` decide what they want to do with the blocks they are given?

So we could have something like this:
```hs
recordPoolCertificates blocks = do
    forM_ blocks $ \case
        BlockByron _ ->
            -- Do nothing, as pools did not exist in the Byron era.
            pure ()
        BlockShelley blk ->
            -- Do stuff.

recordBlockHeaders blocks =
    case NE.last blocks of
        BlockByron blk ->
            putHeader (toByronBlockHeader gp blk)
        BlockShelley blk ->
            putHeader (toShelleyBlockHeader getGenesisBlockHash blk)
```
Then there is no need to invent an extra combinator.

What do you think?

hasufell replied:
As I already explained... I think it's best to explicitly avoid traversing the list twice. This is a hot loop.

> But this function is specialized to a single use case. Having a cryptic name means that the reader has to go to the definition of the function to figure out what it's doing.

Yes it is. I checked if there's something similar already, but I couldn't find it. If you want to suggest a different name, please do so.

> The overall design also seems a bit brittle: if we need to change the way we record block headers (for example), then it's likely we'll need to replace forML_ with something else entirely.

I don't think this is a problem. If the code changes in the future, this will be a simple change.

KtorZ replied:
Not a huge fan of the name `forML_` which is a bit cryptic, but I do agree that traversing the list one once is essential here (that's about 4M elements on mainnet!) 

Perhaps resorting to a slightly more transparent name could help make this clearer. A suggestion: `forAllAndLastM`?

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +557,18 @@ newDBLayer trace fp timeInterpreter = do
+            reverse . fmap (fromBlockHeaders . entityVal) <$> selectList [ ]
+                [ Desc BlockHeight
+                , LimitTo k
+                ]
```
**Request:** add a very simple property test for `listHeaders` that relates it to the `putHeader` and `dropHeaders` functions.

**Justification:**

By adding a property, we can check that both the SQLite implementation and the model implementation behave in the same way (especially if someone else changes this code in future).

In particular, it seems that the ordering is important here. We'd presumably like to know that both the model and the SQLite implementation return the most-recently-added items first.

(We'd presumably also like to know that both implementations respect the `k` parameter.)

For inspiration, see the existing properties in `lib/core/test/unit/Cardano/Pool/DB/Properties.hs`.

hasufell replied:
> In particular, it seems that the ordering is important here. We'd presumably like to know that both the model and the SQLite implementation return the most-recently-added items first.

They return the oldest item first... it's a bit confusing, but the intersection finding function expects them that way.

hasufell replied:
This is the test:

```hs
prop_putHeaderListHeader
    :: DBLayer IO
    -> [BlockHeader]
    -> Positive Int
    -> Property
prop_putHeaderListHeader DBLayer{..} headers (Positive k) =
    monadicIO (setup >> prop)
  where
    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
        . L.nubBy (\a b -> (blockHeight a) == (blockHeight b)) . nub $ headers
    expected = reverse . (if k == 0 then id else take k) . reverse $ sortedHeaders
    setup = run $ atomically cleanDB
    prop = do
        void $ run $ atomically $ forM sortedHeaders putHeader
        lHeaders <- run $ atomically $ listHeaders k
        monitor $ counterexample $ unlines
            [ "Read from DB: " <> show lHeaders
            , "Expected: " <> show expected
            , "Input: " <> show sortedHeaders
            , "k: " <> show k]
        monitor $ classify (null lHeaders) "Empty distributions"
        assert (lHeaders == expected)
        assert (strictlyIncreasing . fmap blockHeight $ lHeaders)

    strictlyIncreasing :: Ord a => [a] -> Bool
    strictlyIncreasing [] = True
    strictlyIncreasing [_] = True
    strictlyIncreasing (x:y:xs) = (x < y) && strictlyIncreasing (y:xs)
```

jonathanknowles reviewed (changes requested):
Hi @hasufell 

Thanks for all your changes after the previous review.

I still have one request and one suggestion:

- **Request**: Add at least a simple property test for the `listBlockHeaders` function.
  See [here](https://github.com/input-output-hk/cardano-wallet/pull/2138#discussion_r489298298) for more details.

- **Suggestion**: Simplify the `forML_` abstraction.
  See [here](https://github.com/input-output-hk/cardano-wallet/pull/2138#discussion_r489294895) for more details.

KtorZ commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -223,6 +223,19 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+        -- ^ List headers, usually stored during syncing
+
+    , dropHeaders
+        :: stm ()
```
May I suggest calling this one `prune` for consistency with the other DBLayer. They have exactly the same purpose.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +557,18 @@ newDBLayer trace fp timeInterpreter = do
+                , LimitTo k
+                ]
+
+        dropHeaders = deleteWhere @_ @_ @TH.BlockHeader []
```
That sounds a little drastic no :) ? It seems like there's a conflict regarding the goal of such function. To me, the dropHeaders / prune function should remove block headers that are no-longer needed (i.e., those that are older than `k`). 

What is the rationale of the current implementation above?

hasufell replied:
I currently have no plans implementing prune. I think the size of the header table is fairly ok, is it not? `dropHeaders` was used previously in the implementation where I kept only one block header in total in the table. This is not the case anymore, so I could remove it altogether.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -837,6 +852,23 @@ fromPoolProduction (PoolProduction pool slot headerH parentH height) =
+    }
+
+fromBlockHeaders :: TH.BlockHeader -> BlockHeader
+fromBlockHeaders TH.BlockHeader { .. } =
```
Note that, whenever possible, we avoid RecordWildCards and prefer instead NamedFieldPuns. RecordWildCards create implicit let bindings in scope that can be hard to follow, especially in large functions. I don't think it's a problem here because the function is straightforward to follow; yet I'd still favor NamedFieldPuns approach for consistency with the rest of the codebase :+1:

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -494,7 +498,9 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
     mkLatestGarbageCollectionEpochRef = newIORef minBound
 
     initCursor :: IO [BlockHeader]
-    initCursor = atomically $ readPoolProductionCursor (max 100 k)
```
I think that the `readPoolProductionCursor` is made obsolete by this PR. So I'd be good to clean it up as well ?

hasufell replied:
It is used in Jormungandr and I haven't touched that code yet.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -505,25 +511,49 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+          where
+            go []  = pure ()
+            go [x] = a1 x >> a2 x >> go []
+            go (x:xs) = a1 x >> go xs
```
:+1:

hasufell commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+prop_putHeaderListHeader DBLayer{..} headers (Positive k) =
+    monadicIO (setup >> prop)
+  where
+    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
```
The generator can't really simulate how we get blocks from the node, so we need to do some deduplicaton and sorting. Pesky.

hasufell commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+  where
+    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
+        . L.nubBy (\a b -> (blockHeight a) == (blockHeight b)) . nub $ headers
+    expected = reverse . (if k == 0 then id else take k) . reverse $ sortedHeaders
```
This simulates the limiting of the results to the latest k elements, while still retaining order of oldest block first.

hasufell commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+            , "k: " <> show k]
+        monitor $ classify (null lHeaders) "Empty distributions"
+        assert (lHeaders == expected)
+        assert (strictlyIncreasing . fmap blockHeight $ lHeaders)
```
I guess this property is kinda implicit in the fact that `expected` is already sorted that way, but I think it's fine to have it explicitly.

jonathanknowles replied:
**Suggestion**: use `assertWith` here, since we have more than one assertion.

**Justification**: when we encounter a failure, it's good to know _which_ assertion failed.

So something like:

```suggestion
        assertWith "read back the headers we expected"
            (lHeaders == expected)
        assertWith "headers are increasing in block height"
            (strictlyIncreasing . fmap blockHeight $ lHeaders)
```

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Gen.hs:
```diff
@@ -153,9 +153,12 @@ genBlockHeader sl = do
+            , Hash $ unsafeFromHex
+                "d93b27cc7bb6fd2fe6ee42de5328c13606bb714a78475a41335207d2afd6026e"
+            , Hash $ unsafeFromHex
+                "63b8828e2eadc3f14b9b691fa9df76139a9c9b13a12ec862b324cc5a88f9fcc5"
```
Hi @hasufell

Could you explain your reasoning for using long hex values here?

AFAIK, the motivation for using short, human-readable strings like "BLOCK0", "BLOCK1" and so on is to make counterexamples more readable. With short strings, we can more easily see the structure of nested values.

hasufell replied:
> AFAIK, the motivation for using short, human-readable strings like "BLOCK0", "BLOCK1" and so on is to make counterexamples more readable.

It failed serialisation/deserialisation.

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+    strictlyIncreasing :: Ord a => [a] -> Bool
+    strictlyIncreasing [] = True
+    strictlyIncreasing [_] = True
+    strictlyIncreasing (x:y:xs) = let !b = (x < y) in b && strictlyIncreasing (y:xs)
```
**Request**: please stick to the [line length limit](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters) if you can!

**Suggestion**: this function could also be expressed as:
```suggestion
    strictlyIncreasing xs = all (uncurry (<)) (xs `zip` tail xs)
```

hasufell replied:
I find this far less readable.

jonathanknowles replied:
That was indeed just a suggestion.

But I would still appreciate it if you could fix the line length issue. This would work:
```suggestion
    strictlyIncreasing [] = True
    strictlyIncreasing [_] = True
    strictlyIncreasing (x:y:xs) =
        let !b = (x < y) in b && strictlyIncreasing (y:xs)
```

hasufell replied:
Done

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+    monadicIO (setup >> prop)
+  where
+    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
+        . L.nubBy (\a b -> (blockHeight a) == (blockHeight b)) . nub $ headers
```
Suggestion: Use [`nubSortOn`](https://hackage.haskell.org/package/extra-1.7.8/docs/Data-List-Extra.html#v:nubSortOn) here:

```suggestion
    sortedHeaders = nubSortOn blockHeight . nub $ headers
```

Also, is the second `nub` actually necessary?

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+  where
+    sortedHeaders = sortBy (\a b -> compare (blockHeight a) (blockHeight b))
+        . L.nubBy (\a b -> (blockHeight a) == (blockHeight b)) . nub $ headers
+    expected = reverse . (if k == 0 then id else take k) . reverse $ sortedHeaders
```
I don't think we need the zero-value test here, as we have already specified that k is `Positive` (which QuickCheck [guarantees as being greater than `0`](http://hackage.haskell.org/package/QuickCheck-2.14.1/docs/Test-QuickCheck-Modifiers.html#t:Positive)).

But even if we did encounter a value of `0`, shouldn't we then just take `0` elements?

I would also be inclined to change the `Positive` to `NonNegative`, as we probably do want to test what happens when `k == 0`.

```suggestion
    expected = reverse . take k . reverse $ sortedHeaders
```

hasufell replied:
> But even if we did encounter a value of 0, shouldn't we then just take 0 elements?

Nope, in that case we should take **all** elements. That's how `LimitTo` from persistent works. This emulates it. And as such, we want the 0-case.

jonathanknowles replied:
> Nope, in that case we should take all elements. That's how `LimitTo` from persistent works. This emulates it. And as such, we want the 0-case.

Is this documented anywhere?

I couldn't find any mention of this in [the Haddock documentation for `LimitTo`](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt).

If this is the case, then we should definitely add a comment in the interface for `DBLayer.listHeaders`, to let callers know this is what they should expect.

hasufell replied:
I don't know, but that's what the implementation does:

https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist.hs#L389-L398

https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist/Sql/Orphan/PersistQuery.hs#L396-L409

jonathanknowles replied:
> I don't know, but that's what the implementation does:

Thanks for confirming.

It's very surprising that this isn't documented anywhere! (I would have thought that the [`LimitTo` constructor](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt) would be an excellent place to put this documentation.)

In any case, I think we should do one of the following:

1. Adjust the SQLite implementation of `listHeaders` to check for `limit <= 0` and then return `[]`.
2. Document the special significance of `limit <= 0` in the interface to `listHeaders`.

My preference would be for option 1, as I'm not a fan of using magic values. IMO, it should _never_ be the case that _reducing_ a limit causes a function to return _more_ values. This breaks the principle of monotonicity.

If OTOH we really need to support the case where caller wants all values, perhaps the limit can be specified as `Maybe Int`, or by using a sum type.

hasufell replied:
I'd propose creating a separate ticket/PR for this. I added documentation about this behavior though.

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1334,3 +1376,7 @@ testCertificatePublicationTimes =
     ]
+
+instance Arbitrary BlockHeader where
+    arbitrary = arbitrary >>= genBlockHeader
```
Suggestion: use `genSlotNo` here:
```suggestion
    arbitrary = genSlotNo >>= genBlockHeader
```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -554,6 +558,17 @@ newDBLayer trace fp timeInterpreter = do
+        putHeader point =
+                let record = mkBlockHeader point
+                    key = TH.blockHeight record
+                in repsert (BlockHeaderKey key) record
```
**Request**: use 4-space indents.

```suggestion
        putHeader point =
            let record = mkBlockHeader point
                key = TH.blockHeight record
            in repsert (BlockHeaderKey key) record
```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -416,6 +422,15 @@ mRemoveRetiredPools epoch = do
+    | k > 0 = reverse . take k <$> get #blockHeaders
+    | otherwise = reverse <$> get #blockHeaders
+
+
```
```suggestion

```
https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Model.hs:
```diff
@@ -416,6 +422,15 @@ mRemoveRetiredPools epoch = do
+mListHeaders :: Int -> ModelOp [BlockHeader]
+mListHeaders k
+    | k > 0 = reverse . take k <$> get #blockHeaders
+    | otherwise = reverse <$> get #blockHeaders
```
Hi @hasufell, is there a good reason for returning the whole list when `k == 0`?

If not, I would be inclined to either remove this special case, or instead write something like:
```hs
mListHeaders :: Int -> ModelOp [BlockHeader]
mListHeaders k
    | k > 0 = reverse . take k <$> get #blockHeaders
    | otherwise = pure []
```

It also seems that this behaviour is different from the SQLite implementation, which just returns 0 rows if `k` is zero. (Though please correct me if I'm wrong here!)

We ideally want the model and SQLite implementations to behave in exactly the same way. (At some point in the future, we'll probably add QSM-style tests that verify this, in the same way we currently do for the wallet DB.)

hasufell replied:
> It also seems that this behaviour is different from the SQLite implementation, which just returns 0 rows if k is zero. (Though please correct me if I'm wrong here!)

I returns all elements if k is 0.

jonathanknowles replied:
> > It also seems that this behaviour is different from the SQLite implementation, which just returns 0 rows if k is zero. (Though please correct me if I'm wrong here!)
> 
> I returns all elements if k is 0.

Is this documented anywhere?

I couldn't find any mention of this in [the Haddock documentation for `LimitTo`](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt).

If this is the case, then we should definitely add a comment in the interface for `DBLayer.listHeaders`, to let callers know this is what they should expect.

jonathanknowles reviewed (changes requested):
Hi @hasufell 

Many thanks for adding the property test!

I still have a few requests and suggestions. (See comments.)

Cheers!

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1294,38 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+prop_putHeaderListHeader
+    :: DBLayer IO
+    -> [BlockHeader]
+    -> Positive Int
```
I would be inclined to change this to `NonNegative`.

Justification: the `listHeaders` function is not just defined on the set of positive values. In particular, it also has defined behaviour when `k == 0`.

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -223,6 +224,17 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
+        :: Int -- ^ limit
+        -> stm [BlockHeader]
+        -- ^ List headers, usually stored during syncing.
+        -- Returns all headers if limit is <= 0.
```
:+1:

jonathanknowles commented on lib/core/src/Cardano/Pool/DB.hs:
```diff
@@ -107,6 +107,7 @@ data DBLayer m = forall stm. (MonadFail stm, MonadIO stm) => DBLayer
         -- be the last element in the list.
         --
         -- This is useful for the @NetworkLayer@ to know how far we have synced.
+        -- Returns all headers if limit is <= 0.
```
:+1:

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1293,56 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
     certificateSequences = getSinglePoolSequences mpcs
     poolCount = length certificateSequences
 
+
```
```suggestion
```

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1293,56 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+  where
+    -- The generator can't really simulate how we get blocks from the node,
+    -- so we need to do some deduplicaton and sorting.
+    sortedHeaders = nubSortOn blockHeight headers
```
:+1: (for use of `nubSortOn`: this is much clearer)

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1293,56 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+    limitTo n
+        | n <= 0 = id
+        -- this emulates persistent [LimitTo n, Desc ...]
+        | otherwise = reverse . take n . reverse
```
:+1:

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1284,6 +1293,56 @@ prop_MultiPoolCertificateSequence_coverage mpcs = checkCoverage
+        -- this emulates persistent [LimitTo n, Desc ...]
+        | otherwise = reverse . take n . reverse
+
+
```
```suggestion
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -702,3 +732,22 @@ instance ToText StakePoolLog where
             , pretty (fixedF 1 (toRational delay / 1000000)), "s"
             ]
+
+
```
```suggestion

```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -702,3 +732,22 @@ instance ToText StakePoolLog where
+        "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
+    (Hash $ unsafeFromHex
+        "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
+
```
```suggestion

```

jonathanknowles commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -131,8 +131,11 @@ import System.FilePath
+import Cardano.Pool.DB.Sqlite.TH hiding
+    ( BlockHeader, blockHeight )
 
+
```
```suggestion
```

jonathanknowles commented on lib/core/test/unit/Cardano/Pool/DB/Properties.hs:
```diff
@@ -1334,3 +1393,7 @@ testCertificatePublicationTimes =
+
+instance Arbitrary BlockHeader where
+    arbitrary = genSlotNo >>= genBlockHeader
+
```
```suggestion
```

jonathanknowles reviewed (approved):
Hi @hasufell 

Thanks for making all the changes.

As for [`LimitTo`](https://hackage.haskell.org/package/persistent-2.10.5.2/docs/Database-Persist-Types.html#t:SelectOpt), I'm still rather shocked that specifying `LimitTo 0` causes `persistent` to issue a query that selects **_all_** rows. IMO, this breaks the principle of monotonicity, in that _decreasing_ a limit should never lead to an _increase_ in the size of the result. (I'm somewhat inclined to file a bug in `persistent` for this surprising behaviour.)

I think it's fine to document the special behaviour for `limit <= 0` in the interface for `DBLayer.listHeaders` for now (as you have done), though I think we should probably try to fix this at some stage so that it really does return `[]`. (Perhaps using `Maybe Int` would work here if we really want to support fetching everything.)

I've approved this PR, but there's still one more thing:

This patch inserts several extraneous carriage returns in a few places. Would you be able to remove the double spaces before merging? I've highlighted these in the comments attached to this review, so you should be able to just hit "apply suggestion" wherever necessary. Thanks! (See [coding standards](https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions).)

KtorZ committed:
remove hard-coded last byron header

  The rationale for removing it is that it's a magic constant which comes a bit out of nowhere and which only works on mainnet. Plus, this only works if the node is already synced (at least after the shelley hard-fork) when starting the wallet. Now that the pool worker makes intermediate checkpoints also in Byron, it can restore a previous state in case of start so there's not much need for jumping right into Shelley, unless maybe for improving performances if that is ever needed.

Commit:
Fix bug in sqlites rollbackTo

This caused infinite recursion in monitorStakePools.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -526,6 +529,7 @@ newDBLayer trace fp timeInterpreter = do
             deleteWhere ([] :: [Filter StakeDistribution])
             deleteWhere ([] :: [Filter PoolMetadata])
             deleteWhere ([] :: [Filter PoolMetadataFetchAttempts])
+            deleteWhere ([] :: [Filter TH.BlockHeader])
```
:ok_hand:

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -479,6 +481,7 @@ newDBLayer trace fp timeInterpreter = do
             deleteWhere [ PoolProductionSlot >. point ]
             deleteWhere [ PoolRegistrationSlot >. point ]
             deleteWhere [ PoolRetirementSlot >. point ]
+            deleteWhere [ BlockSlot >. point ]
```
:ok_hand:
